examples/vision/focal_modulation_network.py

```diff
@@ -0,0 +1,1220 @@
+| :--: |
+| Figure 2: The Entire Architecture (Source: Authors) |
+"""
+
+"""
+# Patch Embedding Layer
+
+The patch embedding layer is used to patchify the input images and project them into a
+latent space. This layer is also used as the down-sampling layer in the architecture.
+"""
+
+
+class PatchEmbed(layers.Layer):
+    """
+    Image patch embedding layer. This layer also acts as the down-sampling layer.
+
+    Parameters:
+        image_size (Tuple[int]): Input image resolution.
+        patch_size (Tuple[int]): Patch spatial resolution.
+        embed_dim (int): Embedding dimnesion.
+        norm_layer (tf.keras.layers.Layer): Normalization layer.
+    """
+
+    def __init__(
+        self,
+        image_size: Tuple[int] = (224, 224),
+        patch_size: Tuple[int] = (4, 4),
+        embed_dim: int = 96,
+        norm_layer: tf.keras.layers.Layer = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        patch_resolution = [
+            image_size[0] // patch_size[0],
+            image_size[1] // patch_size[1],
+        ]
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.embed_dim = embed_dim
+        self.patch_resolution = patch_resolution
+        self.num_patches = patch_resolution[0] * patch_resolution[1]
+        self.proj = layers.Conv2D(
+            filters=embed_dim, kernel_size=patch_size, strides=patch_size
+        )
+        self.flatten = layers.Reshape(target_shape=(-1, embed_dim))
+        self.norm = norm_layer(epsilon=1e-7) if norm_layer is not None else None
+
+    def call(self, x: tf.Tensor) -> Tuple[tf.Tensor, int, int, int]:
+        """
+        Args:
+            x: Tensor of shape (B, H, W, C)
+
+        Returns:
+            A tuple of the processed tensor, height of the projected
+            feature map, width of the projected feature map, number
+            of channels of the projected feature map.
+        """
+        # Project the inputs.
+        x = self.proj(x)
+
+        # Obtain the shape from the projected tensor.
+        H = tf.shape(x)[1]
+        W = tf.shape(x)[2]
+        C = tf.shape(x)[3]
+
+        # B, H, W, C -> B, H*W, C
+        x = self.flatten(x)
+        if self.norm is not None:
+            x = self.norm(x)
+
+        return x, H, W, C
+
+
+"""
+# Focal Modulation _Block_
+
+A Focal Modulation _Block_ can be considered as a single Transformer Block with the Self
+Attention (SA) module being replaced with Focal Modulation module as we saw in **Figure
+2**.
+
+Let us recall how a focal modulation block is supposed to look like with the aid of the
+**Figure 3**.
+
+
+| ![Focal Modulation Block](https://i.imgur.com/bPYTSiB.png) |
+| :--: |
+| Figure 3: The isolated view of the Focal Modulation Block (Source: Authors) |
+
+The Focal Modulation Block consists of:
+- Multilayer Perceptron
+- Focal Modulation _Layer_
+"""
+
+"""
+### Multilayer Perceptron
+"""
+
+
+class MLP(layers.Layer):
```

Comment:
The class-based factoring adds quite a bit of indirection, have you considered doing

```
def MLP(...)
   return Sequential([...])
```

Then you can do

```
y = MLP(...)(x)
```

Reply:
Thank you for this suggestion!